### PR TITLE
SOLR-16040: Fix split packages in hdfs module

### PR DESCRIPTION
**Pull request by risdenk**
https://issues.apache.org/jira/browse/SOLR-16040

Its broken down into multiple commits just as I worked through it. The only weird commit is ce7d256c89eca082a3ebe59595a9d61e098871bc where there were a bunch of changes needed to open up update and transaction classes to allow the new `hdfs` package to work. Everything before was package private so it worked in the same package.

Tested with:
```
./gradlew check
./gradlew test --tests "*hdfs*" --tests "*Hdfs*" --tests "*HDFS*" --tests "*Hadoop*" -Ptests.jvms=4 -Ptests.jvmargs=-XX:TieredStopAtLevel=1 -Ptests.nightly=true -Ptests.awaitsfix=false -Ptests.badapples=false -Ptests.file.encoding=UTF-8
```

### CODE REVIEW

**Comment by dsmiley** [comment]
While you're at it, could you please rename pertinent tests that aren't using camel case for Hdfs like `HDFSCollectionsAPITest` ?  Maybe non-test classes too.  This is the standard I've seen most projects gravitate towards instead of it being haphazard.

File: solr/server/scripts/cloud-scripts/snapshotscli.sh [hunk 1]
```diff
@@ -11,7 +11,7 @@ run_solr_snapshot_tool() {
     log4j_config="file:${scriptDir}/../../resources/log4j2-console.xml"
   fi
   PATH=${JAVA_HOME}/bin:${PATH} ${JVM} ${ZKCLI_JVM_FLAGS} -Dlog4j.configurationFile=${log4j_config} \
-  -classpath "${solrLibPath}" org.apache.solr.core.snapshots.SolrSnapshotsTool "$@" 2> /dev/null
+  -classpath "${solrLibPath}" org.apache.solr.hdfs.snapshots.SolrSnapshotsTool "$@" 2> /dev/null
```
Comment (janhoy) [inline solr/server/scripts/cloud-scripts/snapshotscli.sh:14]: Do we need to add modules/hdfs/... to `solrLibPath` for this tool to work? Why did the SolrSnapshotsTool need to move to hdfs module in the first plae?

**Review by janhoy — COMMENTED** [review]
I love to see how eating our own dogfood improves our base classes by detecting private access for crucial methods and variables, that would prevent users to create their own custom plugins for the same.

Did a quick browse only.

File: solr/server/scripts/cloud-scripts/snapshotscli.sh (hunk 1 above)
Comment (risdenk) [reply solr/server/scripts/cloud-scripts/snapshotscli.sh:14]: `SolrSnapshotsTool` has a bunch of HDFS specific dependencies in it. I looked at trying to separate it a little when moving HDFS to a module, but it wasn't simple. I didn't find any reference to this tool in the ref guide when I checked so I'm not sure how much it is actually used.

**Comment by risdenk** [comment]
@dsmiley I think I fixed all the class names in 1f47ffe

File: solr/server/scripts/cloud-scripts/snapshotscli.sh (hunk 1 above)
Comment (janhoy) [reply solr/server/scripts/cloud-scripts/snapshotscli.sh:14]: The solrLibPath var is set like this
```bash
solrLibPath="${SOLR_LIB_PATH:-${scriptDir}/../../solr-webapp/webapp/WEB-INF/lib/*:${scriptDir}/../../lib/ext/*}"
```
I think `../../../modules/hdfs/lib/solr-hdfs*` is missing there?

**Comment by risdenk** [comment]
@dsmiley fixed `Filesystem` in [f5249ad](https://github.com/apache/solr/pull/688/commits/f5249adf5ae5f284a86c9c8e1490eabcf9782711)

File: solr/server/scripts/cloud-scripts/snapshotscli.sh (hunk 1 above)
Comment (risdenk) [reply solr/server/scripts/cloud-scripts/snapshotscli.sh:14]: Addressed in feec4cbdf31583684d44d1fd78956c6533c28156

Comment (HoustonPutman) [reply solr/server/scripts/cloud-scripts/snapshotscli.sh:14]: The only hdfs specific dependency I see is Path... Is the commons-cli a transative dependency there? In general I'm not sure how useful this tool is now that the backup-restore functionality is really good, but I guess that's for a separate ticket.

File: solr/solr-ref-guide/modules/deployment-guide/pages/backup-restore.adoc [hunk 2]
```diff
@@ -452,7 +452,7 @@ An example configuration using these properties can be found below:
 [source,xml]
 ----
 <backup>
-  <repository name="hdfs" class="org.apache.solr.core.backup.repository.HdfsBackupRepository" default="false">
+  <repository name="hdfs" class="org.apache.solr.hdfs.backup.repository.HdfsBackupRepository" default="false">
```
Comment (HoustonPutman) [inline solr/solr-ref-guide/modules/deployment-guide/pages/backup-restore.adoc:455]: We need to add this in the `major-changes-in-solr-9.adoc`

File: solr/server/scripts/cloud-scripts/snapshotscli.sh (hunk 1 above)
Comment (risdenk) [reply solr/server/scripts/cloud-scripts/snapshotscli.sh:14]: So its not just a compile time dependency on `Path` - if you read the code there are tons of references to HDFS and distcp. Basically for snapshots to work you need a shared filesystem and HDFS is one of the only ones that works semi easily. But yea agree about backup-restore and if snapshots still is needed.

File: solr/solr-ref-guide/modules/deployment-guide/pages/backup-restore.adoc (hunk 2 above)
Comment (risdenk) [reply solr/solr-ref-guide/modules/deployment-guide/pages/backup-restore.adoc:455]: Addressed in [89993fd](https://github.com/apache/solr/pull/688/commits/89993fd056d8cdd8f752805b80968f371f5f2d2c)

File: solr/solr-ref-guide/modules/upgrade-notes/pages/major-changes-in-solr-9.adoc [hunk 3]
```diff
@@ -125,6 +125,8 @@ This is only applicable for users returning information in JSON format, which is
 * SOLR-14660: HDFS storage support has been moved to a module. Existing Solr configurations do not need any HDFS-related
 changes, however the module needs to be installed - see the section xref:deployment-guide:solr-on-hdfs.adoc[].
 
+* SOLR-16040: If you are using the HDFS backup repository, you need to change the repository class to `org.apache.solr.hdfs.backup.repository.HdfsBackupRepository` - see the HDFS section in xref:deployment-guide:backup-restore.adoc[].
```
Comment (HoustonPutman) [inline solr/solr-ref-guide/modules/upgrade-notes/pages/major-changes-in-solr-9.adoc:128]:
```suggestion
* SOLR-16040: If you are using the HDFS backup repository, you need to change the repository class to `org.apache.solr.hdfs.backup.repository.HdfsBackupRepository` - see the xref:deployment-guide:backup-restore.adoc#hdfsbackuprepository[HDFS Backup Repository] section.
```

Comment (HoustonPutman) [inline solr/solr-ref-guide/modules/upgrade-notes/pages/major-changes-in-solr-9.adoc:128]:
```suggestion
* SOLR-16040: If you are using the HDFS backup repository, you need to change the repository class to `org.apache.solr.hdfs.backup.repository.HdfsBackupRepository` - see the xref:deployment-guide:backup-restore.adoc#hdfsbackuprepository[HDFS Backup Repository] section.
```

**Review by HoustonPutman — APPROVED** [review]
This looks good to me with a changelog entry. Though I can't speak for the SolrResourceLoader class.

**Comment by HoustonPutman** [comment]
Did anything else get overridden in the force push?

**Comment by risdenk** [comment]
@HoustonPutman  the force push wasn't the issue. I didn't pull the branch before making local changes to add the `solr/CHANGES.txt`. The only change made through the UI was your documentation change.

**Comment by HoustonPutman** [comment]
Ahh ok, thanks for explaining 🙂

**Pull request merged**